### ISSUE TITLE
Bug: eslint-plugin-import is slow when giant files such as typescript.js are included

### ISSUE THREAD

**Issue by kaiyoma** [import-js/eslint-plugin-import#2416]
I initially filed this bug against eslint (https://github.com/eslint/eslint/issues/15728), but that was analyzed and closed and I was told to file against typescript-eslint instead. I did that (https://github.com/typescript-eslint/typescript-eslint/issues/4735) and after some thorough analysis by the maintainers, they concluded that eslint-plugin-import is the culprit, so posting here.

### Environment

Node version: v16.13.0
npm version: v8.1.4
Local ESLint version: v8.10.0 (Currently used)
Global ESLint version: Not found
Operating System: win32 10.0.19044

### What parser are you using?

@typescript-eslint/parser

### What did you do?

Will post separately (GitHub is complaining about the content length).

### What did you expect to happen?

For eslint to run in a timely manner.

### What actually happened?

eslint is taking almost 2 minutes to lint half a dozen files, none of which are big. In the debug log (posted below), I can see that it's taking over 45 seconds (each) for two particular files to be linted, and neither one is more than 200 lines.

### Additional comments

Will post separately (GitHub is complaining about the content length).

**Comment by kaiyoma** [comment]
For the eslint config, code that will repro the problem, debug logs, and an in-depth analysis of what this plugin is to blame, see the previous issue (https://github.com/typescript-eslint/typescript-eslint/issues/4735).

**Comment by burtek** [comment]
Same here, working on an eslint plugin, imported `typescript` and it takes ages to lint a single file:

```
$ eslint src/eslint-utils/isMethodCalledOnArray.ts --fix-dry-run

[REDACTED]/src/eslint-utils/isMethodCalledOnArray.ts
  10:23  warning  No magic number: 1  @typescript-eslint/no-magic-numbers
  13:23  warning  No magic number: 2  @typescript-eslint/no-magic-numbers
  16:23  warning  No magic number: 3  @typescript-eslint/no-magic-numbers

✖ 3 problems (0 errors, 3 warnings)

Rule                                   | Time (ms) | Relative
:--------------------------------------|----------:|--------:
import/no-deprecated                   | 27330.375 |    38.6%
import/namespace                       | 25952.951 |    36.6%
import/no-cycle                        | 17408.762 |    24.6%
@typescript-eslint/naming-convention   |    42.714 |     0.1%
n/no-extraneous-import                 |    13.045 |     0.0%
n/no-unpublished-import                |     9.993 |     0.0%
@typescript-eslint/indent              |     9.372 |     0.0%
@typescript-eslint/no-misused-promises |     7.999 |     0.0%
import/order                           |     5.278 |     0.0%
import/no-duplicates                   |     4.737 |     0.0%
Done in 74.57s.
```

**Comment by burtek** [comment]
@JounQin give me a few minutes, will let you know

**Comment by burtek** [comment]
@JounQin it is faster, but not much (I only replaced the 3 longest taking rules as I'm using shared config so it'd be hard to replicate all of it and replace `import` with `i`)

```
Rule                                   | Time (ms) | Relative
:--------------------------------------|----------:|--------:
i/no-deprecated                        | 19176.082 |    43.6%
i/namespace                            | 16235.055 |    36.9%
i/no-cycle                             |  8471.943 |    19.2%
@typescript-eslint/naming-convention   |    37.586 |     0.1%
n/no-extraneous-import                 |     8.199 |     0.0%
@typescript-eslint/indent              |     7.042 |     0.0%
@typescript-eslint/no-misused-promises |     6.244 |     0.0%
n/no-unpublished-import                |     5.568 |     0.0%
import/order                           |     2.679 |     0.0%
import/no-duplicates                   |     1.913 |     0.0%
Done in 48.12s.
```

**Comment by JounQin** [comment]
@burtek 

```sh
npm install -D eslint-plugin-import@npm:eslint-plugin-i
```

It should not needed to change your eslint config at all.

**Comment by burtek** [comment]
@JounQin oh, didn't know that's possible 😮

**Comment by JounQin** [comment]
> @JounQin oh, didn't know that's possible 😮

That's documented at https://github.com/un-es/eslint-plugin-i#installation 😏

**Comment by burtek** [comment]
@JounQin who reads docs nowadays 🙃 😆
